sections/Chapter 7 Sprint 5/virtuoso_part_2.tex

```diff
@@ -0,0 +1,115 @@
+
+Another potential choice could be Linked Data Fragments (LDF). However, LDF seems to use HDT, which would not solve the issue of restarting the database periodically. Therefore, Virtuoso remains our choice.
+
+When we first looked at Virtuoso, we found some frameworks that would help us facilitate data insertion and retrieval.  However, during implementation, none of these frameworks proved useful. Multiple frameworks, like \textit{dotNetRDF} (C\#) and \textit{RDFLib} (Python), were attempted to be used, in order to reduce complexity of the implementation. However, we were unable to establish a connection to the database with any of the frameworks, so the efforts were abandoned. Instead, we decided to use the Virtuoso SPARQL endpoint directly, which worked well for our needs.
+
+As for our implementation, we did not choose to publish the SPARQL endpoint for \knox{} usage directly. Instead, we set up a web API in C\#.
+Our reason for doing this is to encapsulate database access, such that we can change the database in the future without affecting the API.
+This API is connected to Virtuoso, and both run in a dockerized environment. They run on the same network, so that they can communicate with each other.
+
+As mentioned, Virtuoso supports insertion of data in Turtle format, and retrieval of data in RDF/XML format. Inserting and querying data is done through the Virtuoso SPARQL endpoint. The C\# API we have set up relies on this endpoint to do the actual work. We pass on the queries and data, acting as a proxy between the user and the Virtuoso endpoint. At least, this is the case when users send SPARQL queries to the API. When users send Turtle data to the API, we wrap it in a SPARQL query with some preset parameters, and then send it to Virtuoso. For example, we have specified a default graph, in case the user does not specify one.
+This can be seen in the \texttt{Insert} function, shown in code snippet \ref{lst:virtuoso_data_insertion}. Users send a HTTP POST request to the API, with the data in Turtle format as a string, along with a string denoting the graph name they want to insert the data to. The API then sends the data to Virtuoso, which can be seen in \texttt{InsertTurtleGraph}, also shown in code snippet \ref{lst:virtuoso_data_insertion}. As can be seen in the definition for \texttt{InsertTurtleGraph}, we have specified a default graph. First, we parse the Turtle data from the input string, so we are certain that it is valid Turtle. Then, we log how many triples we have found in the given graph, after which we chunk and send the data to Virtuoso. We will return to the chunking logic momentarily, but for now, know that it returns the Turtle data as an array of strings, each of which is a chunk of data. Each chunk is wrapped in a SPARQL query that specifies that data should be inserted into the given graph. This query is then sent to the Virtuoso endpoint through HTTP.
```

Comment:
```suggestion
This can be seen in the \texttt{Insert} function, shown in code snippet \ref{lst:virtuoso_data_insertion}. Users send an HTTP POST request to the API, with the data in Turtle format as a string, along with a string denoting the graph name they want to insert the data to. The API then sends the data to Virtuoso, which can be seen in \texttt{InsertTurtleGraph}, also shown in code snippet \ref{lst:virtuoso_data_insertion}. As can be seen in the definition for \texttt{InsertTurtleGraph}, we have specified a default graph. First, we parse the Turtle data from the input string, so we are certain that it is valid Turtle. Then, we log how many triples we have found in the given graph, after which we chunk and send the data to Virtuoso.  Each chunk is wrapped in a SPARQL query that specifies that data should be inserted into the given graph. This query is then sent to the Virtuoso endpoint through HTTP.
```

Reply:
Why remove that sentence?

Reply:
It's very 'talkative' and did not add new information. Essentially, you say "don't worry about it" and then explain how it is done. It also breaks the reading flow.